kani-compiler/src/kani_middle/transform/loop_contracts.rs

```diff
@@ -0,0 +1,266 @@
+// Copyright Kani Contributors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+//! This module contains code related to the MIR-to-MIR pass to enable loop contracts.
+//!
+
+use crate::kani_middle::codegen_units::CodegenUnit;
+use crate::kani_middle::find_fn_def;
+use crate::kani_middle::transform::body::{MutableBody, SourceInstruction};
+use crate::kani_middle::transform::{TransformPass, TransformationType};
+use crate::kani_middle::KaniAttributes;
+use crate::kani_queries::QueryDb;
+use crate::stable_mir::CrateDef;
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Symbol;
+use stable_mir::mir::mono::Instance;
+use stable_mir::mir::{BasicBlockIdx, Body, Operand, Terminator, TerminatorKind};
+use stable_mir::ty::{FnDef, RigidTy};
+use stable_mir::DefId;
+use std::collections::VecDeque;
+use std::collections::{HashMap, HashSet};
+use std::fmt::Debug;
+use tracing::trace;
+
+/// This pass will perform the following operations:
+/// 1. Replace the body of `kani_register_loop_contract` by `kani::internal::run_contract_fn`
+///    to invoke the closure.
+///
+/// 2. Replace the dummy call to the register function with the actual call, i.e., transform
+///
+/// ```ignore
+/// let kani_loop_invariant = || -> bool {inv};
+/// kani_register_loop_contract(kani_loop_invariant)
+/// while guard {
+///     loop_body;
+///     kani_register_loop_contract(||->bool{true});
```

Comment:
wait, why do you register a dummy closure down here?

I think we want to redefine the closure so it captures the arguments again and don't violate the borrow checker. Something like:

```rust
let kani_loop_invariant_entry = || -> bool {inv};
kani_register_loop_contract(kani_loop_invariant_entry)
while guard {
    loop_body;
    let kani_loop_invariant_latch = || -> bool {inv};
    kani_register_loop_contract(kani_loop_invariant_latch);
}
```

Reply:
The idea of this implementation (and also the loop contracts in CBMC) is that we want to use one single expression as the loop invariants, and let CBMC do the transformation:  rewrite loop to if, duplicate the loop invariant (check and induction assumption) to the entry of the loop as base case, instrument havocing instructions after the induction assumption, and check the loop invariant at the end of the loop.

The one single loop invariant expression in my mind should be `kani_loop_invariant` which is `|| -> bool {inv}`. As it will be called before the entry of the loop after the transformation, it should be assigned before the loop. However, without dummy call, expanding the macro in the following way will violate the borrow check.
```
let kani_loop_invariant = || -> bool {inv};
while guard {
    loop_body;
    kani_register_loop_contract(kani_loop_invariant);
}
```

So I choose to use a dummy call as a place holder and later replace it with the actual call during MIR transform.

Reply:
But this MIR transformation is creating an invalid MIR that does not respect borrow checking rules.